library/common/http/dispatcher.cc

```diff
@@ -34,50 +34,59 @@ void Dispatcher::DirectStreamCallbacks::encodeHeaders(const ResponseHeaderMap& h
   ENVOY_LOG(debug, "[S{}] response headers for stream (end_stream={}):\n{}",
             direct_stream_.stream_handle_, end_stream, headers);
 
-  uint64_t response_status = Utility::getResponseStatus(headers);
+  ASSERT(http_dispatcher_.getStream(direct_stream_.stream_handle_));
 
-  // Record received status to report success or failure stat in
-  // Dispatcher::DirectStreamCallbacks::closeRemote. Not reported here to avoid more complexity with
-  // checking end stream, and the potential for stream reset, thus resulting in mis-reporting.
-  observed_success_ = CodeUtility::is2xx(response_status);
+  uint64_t response_status = Utility::getResponseStatus(headers);
+  // Track success for later bookkeeping (stream could still be reset).
+  success_ = CodeUtility::is2xx(response_status);
 
-  // @see Dispatcher::resetStream's comment on its call to runResetCallbacks.
-  // Envoy only deferredDeletes the ActiveStream if it was fully closed otherwise it issues a reset.
-  if (direct_stream_.local_closed_ && end_stream) {
-    direct_stream_.hcm_stream_pending_destroy_ = true;
+  if (end_stream) {
+    closeStream();
   }
 
   // TODO: ***HACK*** currently Envoy sends local replies in cases where an error ought to be
   // surfaced via the error path. There are ways we can clean up Envoy's local reply path to
   // make this possible, but nothing expedient. For the immediate term this is our only real
   // option. See https://github.com/lyft/envoy-mobile/issues/460
 
-  // The presence of EnvoyUpstreamServiceTime implies these headers are not due to a local reply.
-  if (headers.get(Headers::get().EnvoyUpstreamServiceTime) != nullptr) {
-    // Testing hook.
-    http_dispatcher_.synchronizer_.syncPoint("dispatch_encode_headers");
-
-    if (direct_stream_.dispatchable(end_stream)) {
-      ENVOY_LOG(debug,
-                "[S{}] dispatching to platform response headers for stream (end_stream={}):\n{}",
-                direct_stream_.stream_handle_, end_stream, headers);
-      bridge_callbacks_.on_headers(Utility::toBridgeHeaders(headers), end_stream,
-                                   bridge_callbacks_.context);
-      closeRemote(end_stream);
+  // Error path: missing EnvoyUpstreamServiceTime implies this is a local reply, which we treat as
+  // a stream error.
+  if (!success_ && headers.get(Headers::get().EnvoyUpstreamServiceTime) == nullptr) {
+    ENVOY_LOG(debug, "[S{}] intercepted local response", direct_stream_.stream_handle_);
+    mapLocalResponseToError(headers, end_stream);
+    if (end_stream) {
+      onReset();
```

Comment:
looks like in the 200 case we are firing `onComplete()` in `mapLocalResponseToError` and then `onRese()` here?

Reply:
Actually the 200 case will never match because I check success first - I just forgot to delete it.